content/blog/2020-03-dplyr-1-0-0-new-verb-features/index.Rmarkdown

```diff
@@ -0,0 +1,231 @@
+
+### Quantiles
+
+To demonstrate this new feature we'll start by looking at a summary that was previous hard to compute with `summarise()`. `quantile()` was hard to use because it returns multiple values which used to cause `summarise()` to error. Now it's straightforward:
+
+```{r}
+df <- tibble(
+  grp = rep(1:2, each = 10), 
+  x = c(rnorm(10, -0.25, 1), rnorm(10, 0, 1.5)),
+  y = c(rnorm(10, 0.25, 1), rnorm(10, 0, 0.5)),
+)
+df %>% 
+  group_by(grp) %>% 
+  summarise(x = quantile(x, c(0.25, 0.5, 0.75)), q = c(0.25, 0.5, 0.75))
+```
+
+It would be nice to be able to reduce the duplication in this code so that we don't have to type the quantile values twice. We can now write a simple function to do so because `summarise()` expressions can now return multiple columns:
+
+```{r}
+quantile2 <- function(x, q = c(0.25, 0.5, 0.75)) {
+  tibble(x = quantile(x, q), q = q)
+}
+df %>% 
+  group_by(grp) %>% 
+  summarise(quantile2(x, c(0.25, 0.5, 0.75)))
+```
+
+In the past, one of the challenges of writing this sort of function was naming the columns. For example, when you call `quantile2(y)` it'd be nice if you'd get columns `y` and `y_q`, not `x` and `x_q`. Now, thanks to the recent combination of [glue and tidy evaluation](https://www.tidyverse.org/blog/2020/02/glue-strings-and-tidy-eval/) that behaviour is straightforward to implement: 
+
+```{r}
+quantile2 <- function(x, q = c(0.25, 0.5, 0.75)) {
+  tibble("{{ x }}" := quantile(x, q), "{{ x }}_q" := q)
+}
+
+df %>% 
+  group_by(grp) %>% 
+  summarise(quantile2(y, c(0.25, 0.5, 0.75)))
+```
+Figuring out how to name the output columns is a surprisingly complex task and we're still thinking about the best approach. 
+
+### Packed columns
+
+Note that in the code above, we've been careful not to name the result of `quantile2()`. When we do that, the data frame result is automatically **unpacked** so each column becomes a column in the result. What happens if we name the output?
+
+```{r}
+out <- df %>% 
+  group_by(grp) %>% 
+  summarise(y = quantile2(y, c(0.25, 0.75)))
+out
+```
+Look carefully at the output - you'll see a `$` in the column names. This is a suggestion that something weird is going on and you have what we call a **df-column** because you have a column of a data frame that is itself a data frame! You can confirm that by extracting just that column:
+
+```{r}
+out$y
+```
+
+And of course, you can dig still deeper to get the individual values:
+
+```{r}
+out$y$y
+```
+
+Df-columns are surprisingly important to the internals of dplyr 1.0.0, but you should able to ignore their existence unless you deliberately want to try them out. They're definitely an advanced topic, and are something that we're continuing to play around with. We're also thinking about how improve the tibble print method to make it more obvious that something unusual is going on.
```

Comment:
If df-cols section remains public facing ... is it worth making a vague connection to spreadsheets/tables with merged column headers?